OpenTrons/opentrons-modules
# feat(heater-shaker): Add bootloader entry

We would like to be able to update heater-shaker units from the field. There is a built-in bootloader on these parts ([AN2606](https://www.st.com/resource/en/application_note/cd00167594-stm32-microcontroller-system-memory-boot-mode-stmicroelectronics.pdf) is the relevant datasheet on the bootloaders). It exposes something not entirely unlike standard DFU. It also has some bizarre entry methods on the F303 part - you must change the state of an external pin, which would require extra circuitry.

However, the machinery that maps different parts of memory to 0x0 is user accessible in syscfg, and we know how cortex-m boot works, so we can jump to that built-in bootloader  from software after making sure we reset our clocking to something that it expects.

This has to be done in assembler because a c call messes around with the stack, which we don't want, so it's a little ugly, but it does work.

We also add the `dfu` command, which will enter this mode after stopping any ongoing heater or motor commands and nicely disconnecting from USB.

## Testing
- Connect a board via USB
- send it "dfu"
- see that  it goes away and comes back reporting itself as an stm32 bootloader
- use http://dfu-util.sourceforge.net/ or something similar for non-osx platforms to mess around with it. You can tell it to boot back to application mode using the `-s 0x0800000:leave` argument  to dfu-util. You can program  it from here but you have to put it in an annoying format first which I'll get to later

Sorry about the assembly.
